perf_bench_test.go

```diff
@@ -24,127 +24,241 @@ package main
+
+	if isRunPerf {
+		result, err := perfStat.GetResult()
+		require.NoError(t, err, "Stop perf stat returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f\n", eventName, value)
+			serveMetrics[eventName] = value
+		}
+	}
+	log.Debugf("average-execution-time: %f\n", serveMetrics[averageExecutionTime])
+	log.Debugf("real-requests-per-second: %f\n", serveMetrics[realRequestsPerSecond])
+
+	writeResultToCSV(t, "benchRPS.csv", serveMetrics)
 }
 
-///////////////////////////////////////////////////////////////////////////////
-////////////////////////// Auxialiary functions below /////////////////////////
-///////////////////////////////////////////////////////////////////////////////
-func serveVM(t *testing.T, start time.Time, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool) {
+func serveVM(t *testing.T, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool, avgExecTime, realRPS *int64) {
 	defer vmGroup.Done()
 
+	tStart := time.Now()
 	resp, _, err := funcPool.Serve(context.Background(), vmIDString, imageName, "replay")
-	require.NoError(t, err, "Function returned error")
-	require.Equal(t, resp.Payload, "Hello, replay_response!")
-
-	if *isColdStart {
-		require.Equal(t, resp.IsColdStart, true)
-		message, err := funcPool.RemoveInstance(vmIDString, imageName, isSyncOffload)
-		require.NoError(t, err, "Function returned error, "+message)
+	require.Equal(t, resp.IsColdStart, false)
+	if err != nil {
+		log.Warn("Function returned error")
 	} else {
-		require.Equal(t, resp.IsColdStart, false)
+		if resp.Payload != "Hello, replay_response!" {
+			log.Warnf("Function returned invalid: %s", resp.Payload)
+		}
+		atomic.AddInt64(realRPS, 1)
+	}
+
+	execTime := time.Since(tStart).Milliseconds()
+	atomic.AddInt64(avgExecTime, execTime)
+	log.Debugf("VM %s: returned in %d milliseconds", vmIDString, execTime)
+}
+
+// Returns a list of image names
+func getImages(t *testing.T) []string {
+	var (
+		images = map[string]string{
+			"helloworld":   "vhiveease/helloworld:var_workload",
+			"chameleon":    "vhiveease/chameleon:var_workload",
+			"pyaes":        "vhiveease/pyaes:var_workload",
+			"image_rotate": "vhiveease/image_rotate:var_workload",
+			"json_serdes":  "vhiveease/json_serdes:var_workload",
+			"lr_serving":   "vhiveease/lr_serving:var_workload",
+			"cnn_serving":  "vhiveease/cnn_serving:var_workload",
+			"rnn_serving":  "vhiveease/rnn_serving:var_workload",
+		}
+		funcs  = strings.Split(*funcNames, ",")
+		result []string
+	)
+
+	for _, funcName := range funcs {
+		imageName, isPresent := images[funcName]
+		require.True(t, isPresent, "Function %s is not supported", funcName)
+		result = append(result, imageName)
+	}
+
+	return result
+}
+
+func writeResultToCSV(t *testing.T, outfile string, metrics map[string]float64) {
+	outFile := getOutFile(outfile)
+
+	f, err := os.OpenFile(outFile, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
```

Comment:
correct the permissions to 666 so that any user can delete this file. (if there are any other files that your code produces, they all should have these permissions; the folders must be 777 for the same reason)

Reply:
Done